frame/dapps-staking/src/tests.rs

```diff
@@ -1,29 +1,57 @@
 use super::{Event, *};
+use frame_support::{assert_err, assert_noop, assert_ok};
 use mock::*;
 
-use frame_support::{assert_noop, assert_ok};
-
 #[test]
-fn assert_is_ok() {
+fn bonding_is_ok() {
     ExternalityBuilder::build().execute_with(|| {
         let stash_id = Origin::signed(1);
         let controller_id = 2u64;
 
-        assert_ok!(DappStaking::bond(
+        assert_ok!(DappsStaking::bond(
             stash_id,
             controller_id,
             50,
             crate::RewardDestination::Staked
         ));
-        // let bond_event = Event::pallet_dapps_staking(RawEvent::Bonded(stash_id, 50));
-        println!("{:?}", System::events()[0].event);
-        assert_eq!(get_dapp_staking_events()[0], Event::Bonded(1, 50));
+        System::assert_last_event(mock::Event::DappsStaking(crate::Event::Bonded(1, 50)));
     })
 }
 
-// pub fn bond(
-//     origin: OriginFor<T>,
-//     controller: <T::Lookup as StaticLookup>::Source,
-//     #[pallet::compact] value: BalanceOf<T>,
-//     payee: RewardDestination<T::AccountId>,
-// ) -> DispatchResultWithPostInfo {
+#[test]
+fn bonding_is_not_ok() {
+    ExternalityBuilder::build().execute_with(|| {
+        let stash_id = Origin::signed(1);
+        let stash2_id = Origin::signed(2);
+        let controller_id = 3u64;
+
+        assert_ok!(DappsStaking::bond(
+            stash_id.clone(),
+            controller_id.clone(),
+            50,
+            crate::RewardDestination::Staked
+        ));
+
+        // repeat bonding with same stash account, expect error AlreadyBonded
+        assert_err!(
+            DappsStaking::bond(
+                stash_id,
+                controller_id.clone(),
+                50,
+                crate::RewardDestination::Staked
+            ),
+            crate::pallet::pallet::Error::<TestRuntime>::AlreadyBonded
+        );
+
+        // use already paired controller, expect error AlreadyPaired
+        assert_err!(
+            DappsStaking::bond(
+                stash2_id,
+                controller_id,
+                50,
+                crate::RewardDestination::Staked
+            ),
+            crate::pallet::pallet::Error::<TestRuntime>::AlreadyPaired
+        );
```

Comment:
You don't cover the scenario where there are insufficient funds (below existential deposit).

Is this even possible if we don't properly mock pallet_balances?

In general, it seems very strange that we're using real implementation of pallet_balances to test pallet for dapp staking.

Reply:
Nvm, disregard the comment.